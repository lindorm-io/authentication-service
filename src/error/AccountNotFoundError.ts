import { APIError, HttpStatus } from "@lindorm-io/core";

export class AccountNotFoundError extends APIError {
  constructor(accountId?: string) {
    super("Account not found", {
      debug: { accountId },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
