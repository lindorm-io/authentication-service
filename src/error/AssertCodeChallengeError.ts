import { APIError, HttpStatus } from "@lindorm-io/core";

export class AssertCodeChallengeError extends APIError {
  constructor(challenge: string, verifier: string) {
    super("Invalid Code Challenge", {
      debug: { challenge, verifier },
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
