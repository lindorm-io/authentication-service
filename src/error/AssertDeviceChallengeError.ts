import { APIError } from "@lindorm-io/errors";
import { HttpStatus } from "@lindorm-io/core";

export class AssertDeviceChallengeError extends APIError {
  constructor(challenge: string, verifier: string) {
    super("Invalid Device Challenge", {
      debug: { challenge, verifier },
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
