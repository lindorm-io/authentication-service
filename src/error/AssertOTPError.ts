import { APIError, HttpStatus } from "@lindorm-io/core";

export class AssertOTPError extends APIError {
  constructor() {
    super("Invalid OTP", {
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
