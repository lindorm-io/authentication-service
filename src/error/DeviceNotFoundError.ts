import { APIError, HttpStatus } from "@lindorm-io/core";

export class DeviceNotFoundError extends APIError {
  constructor(deviceId?: string) {
    super("Device not found", {
      debug: { deviceId },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
