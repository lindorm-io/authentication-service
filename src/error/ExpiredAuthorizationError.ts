import { APIError } from "@lindorm-io/errors";
import { HttpStatus } from "@lindorm-io/core";
import { Authorization } from "../entity";

export class ExpiredAuthorizationError extends APIError {
  constructor(authorization: Authorization) {
    super("Authorization is expired", {
      debug: {
        id: authorization.id,
        expires: authorization.expires,
      },
      statusCode: HttpStatus.ClientError.UNAUTHORIZED,
    });
  }
}
