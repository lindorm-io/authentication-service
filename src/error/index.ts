export * from "./AccountNotFoundError";
export * from "./AssertClientSecretError";
export * from "./AssertCodeChallengeError";
export * from "./AssertDeviceChallengeError";
export * from "./AssertOTPError";
export * from "./DeviceNotFoundError";
export * from "./ExpiredSessionError";
export * from "./InvalidAuthorizationError";
export * from "./InvalidAuthorizationTokenError";
export * from "./InvalidClientError";
export * from "./InvalidDeviceError";
export * from "./InvalidExpiryString";
export * from "./InvalidGrantTypeError";
export * from "./InvalidPermissionError";
export * from "./InvalidRefreshTokenError";
export * from "./InvalidScopeError";
export * from "./InvalidSubjectError";
export * from "./LockedAccountError";
export * from "./RejectedClientError";
export * from "./RequestLimitBackOffError";
