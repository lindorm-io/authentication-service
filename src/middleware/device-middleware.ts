import { IAuthContext } from "../typing";
import { InvalidDeviceError } from "../error";
import { TPromise } from "@lindorm-io/core";
import Joi from "@hapi/joi";

export interface IDeviceMiddlewareOptions {
  throwError: boolean;
}

const schema = Joi.object({
  deviceId: Joi.string().guid().required(),
});

export const deviceMiddleware = (options: IDeviceMiddlewareOptions) => async (
  ctx: IAuthContext,
  next: TPromise<void>,
): Promise<void> => {
  const start = Date.now();

  const { logger, repository } = ctx;
  const { throwError } = options;
  const { deviceId } = ctx.request.body;

  try {
    await schema.validateAsync({ deviceId });

    ctx.device = await repository.device.find({ id: deviceId });

    logger.debug("device found", { deviceId });
  } catch (err) {
    if (throwError) {
      throw new InvalidDeviceError(deviceId, err);
    }

    logger.debug("device not found", { deviceId });
  }

  ctx.metrics = {
    ...(ctx.metrics || {}),
    device: Date.now() - start,
  };

  await next();
};
