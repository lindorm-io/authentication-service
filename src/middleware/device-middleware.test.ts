import MockDate from "mockdate";
import { InvalidDeviceError } from "../error";
import { MOCK_LOGGER, MOCK_UUID, getMockRepository } from "../test/mocks";
import { deviceMiddleware } from "./device-middleware";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("deviceMiddleware", () => {
  let getMockContext: any;
  let next: any;

  beforeEach(() => {
    getMockContext = () => ({
      logger: MOCK_LOGGER,
      repository: getMockRepository(),
      request: { body: { deviceId: MOCK_UUID } },
    });
    next = () => Promise.resolve();
  });

  test("should successfully set device on ctx", async () => {
    const ctx = getMockContext();

    await expect(deviceMiddleware({ throwError: true })(ctx, next)).resolves.toBe(undefined);

    expect(ctx.device).toMatchSnapshot();
    expect(ctx.metrics.device).toStrictEqual(expect.any(Number));
  });

  test("should skip setting device when there is no device id", async () => {
    const context = getMockContext();
    const ctx = {
      ...context,
      repository: {
        ...context.repository,
        device: {
          ...context.repository.device,
          find: jest.fn(() => {
            throw new Error();
          }),
        },
      },
    };

    await expect(deviceMiddleware({ throwError: false })(ctx, next)).resolves.toBe(undefined);

    expect(ctx.device).toBe(undefined);
  });

  test("should throw error on missing device", async () => {
    const context = getMockContext();
    const ctx = {
      ...context,
      repository: {
        ...context.repository,
        device: {
          ...context.repository.device,
          find: jest.fn(() => {
            throw new Error();
          }),
        },
      },
    };

    await expect(deviceMiddleware({ throwError: true })(ctx, next)).rejects.toStrictEqual(
      expect.any(InvalidDeviceError),
    );
  });
});
