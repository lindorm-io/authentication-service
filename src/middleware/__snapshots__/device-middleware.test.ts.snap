// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deviceMiddleware should successfully set device on ctx 1`] = `
Device {
  "_accountId": "accountId",
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": null,
  "_pin": Object {
    "signature": null,
    "updated": null,
  },
  "_publicKey": "-----BEGIN RSA PUBLIC KEY-----
MIICCgKCAgEAs7ydT9Ndt1Idx9NE2zMhAcecJGUcIlbWeXEtZJA7aKZDDhxYaEGB
WZvX+GRndllnPrOgzoxukITGYvj88bnrGZVZX2ckjIVUR1NxDrhyMjqnn4yDz4UB
Q3Zqt6Gz6DcE+Y9ro6kyhu01KPMyUlhDQAUdwpIJA3saUJ5KZj6ai4NTuXbLqzly
+3H5TMQY4WjrQwj/KRuq8LBEsR4YGaiaFZjHaf5n8+s2Gim/dDgAbIsZS+lGTwgx
nigd7dKP9tG6q3GwNPNBSuuQh6rVbBKbmzBa5rO9SmeZ+HLUyClpzhZ+Hp8yOGAn
qQFZZnokJ6dhxDEQcxPV2HUsYpi3oYSSomXsLgYSja1VT7CgeBchWj07ux6MnagM
AB1+O7O2yzu0LYJc72BQOJaSIrBCh1FcPjM0KC1CPocux2JtqgDL/Td+V5HbLU9n
ZQQoYDcrK3ozSne9ojioUWEwdBEzQmj9e3Wei87EyLe5bn/6tDbg17R7ZlVy2iVs
IFziLHrv3VUIbESy5tTCdzK/qzFR0hlxejT12CTUSkN5eKmfz0pfzt+w8ldEfpws
lBUcoS4sBIR3urpvx1EKM1xaUHfchhIKnY5KGTEM+5rJVwlDq+3lLcrJI3AErTAX
gSR6fcZmKuI4nCfTKl3ykozcPwpEmfRgSBDHKwDJe6ZbK0sL4ScVookCAwEAAQ==
-----END RSA PUBLIC KEY-----
",
  "_secret": null,
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
